### ISSUE TITLE
Run current scene (F6) often launches unexpected scene

### ISSUE THREAD

**Issue by h0lley** [godotengine/godot#75998]
### Godot version

v4.0.2.stable.official [7a0977ce2]

### System information

Ubuntu

### Issue description

I would expect F6 to always launch the scene that is currently selected from the scene tab bar. This also seems to have been the case up until recently.

However, currently it seems that if test play is already running, and F6 is used to not just launch a new test play process but also terminate the previous one, then it always loads the same scene the previous test play session has been started with - regardless of the scene tab that is actually selected in the editor.

So a workaround appears to be to always manually stop test play (F8) before starting it again, but this messes a lot with the UX and workflow of rapid prototyping.

### Steps to reproduce

1. Start any project with at least two scenes
2. Launch into scene A via F6, keep it running in the background
3. Then in the editor, open scene B
4. Press F6 again to launch into scene B
5. Note that test play unexpectedly starts with scene A again

### Minimal reproduction project

should be trivial to reproduce in any project, let me know if not

**Comment by Calinou** [comment]
I remember seeing this reported in another issue, but I can't find it right now.

**Comment by h0lley** [comment]
> This may actually be intended behavior though – pressing F6 without stopping the project first will restart the project, not run the current scene.

that is not the behavior of Godot 3, and I am pretty sure it changed in 4 only a couple of weeks or perhaps roughly two months ago at most (that is roughly when I started noticing and being annoyed by it, but only today it triggered me sufficiently to make me look into reproduction 😅).

plus I don't see how that would make any intuitive sense nor is it being communicated in the tooltip. F6 should have only one function and not a hidden second one. restarting the project can still be done by pressing F5 or F6 depending on with which scene you want to start with the original behavior.

**Comment by AThousandShips** [comment]
Nope this changed some 8 months ago

**Comment by h0lley** [comment]
ah, interesting.
I actually never realized that when starting test play, the `Run current scene (F6): Play the edited scene` button turns into a `Run current scene (F6): Reload the played scene` button.

I think this is confusing and don't see the usability in it.
the behavior of Godot 3 makes is what I'd expect intuitively and has greater usability.

|Behavior|Action|Use case|
|Godot 3|Pressing F5||
|Godot 3|Pressing F5 when the game already runs||
|Godot 4|Pressing F6||

**Comment by h0lley** [comment]
ah, interesting.
I actually never realized that when starting test play, the `Run current scene (F6): Play the edited scene` button turns into a `Run current scene (F6): Reload the played scene` button.

I think this is confusing and don't see the usability in it.
the behavior of Godot 3 is what I'd expect intuitively and has greater usability.

I would see some use case in the current behavior if it provided a way to reload the `current_scene` of the running game (for instance, after the scene has been changed via `get_tree().change_scene_to_*`, it would still remain on that scene), but that is not the case; it just always starts the game with the scene it has been previously started with. but even then, that would make more sense as a dedicated button instead of overwriting the `Run current scene (F6)` function.

if F6 would just always launch the scene that is currently selected in the editor (just like is the case in Godot 3), then the user would have the greatest control over what scene is to be launched with.

**Comment by YuriSizov** [comment]
> also, I just had a look at the report (#31515) of the user for whom this has been changed for, and to me it sounds like they were also not asking for the current behavior, [but were expecting](https://github.com/godotengine/godot/issues/31515#issuecomment-523381259) that the reload button would basically trigger `get_tree().reload_current_scene()`.

The comment doesn't say that. The current behavior fits that description: the button reloads currently running scene. It does so by restarting a debug session, but so what? 🙃 It's a quick and clear solution to this problem that makes sure there are no side-effects.

**Comment by AThousandShips** [comment]
It running the scene you last started makes more sense if it isn't just run the currently open scene, for example you may often have scenes that depend on being started from another scene and lack the functionality to be started directly at least without weird issues (and the tooltip clarifies this)

**Comment by YuriSizov** [comment]
> no, it restarts with the scene the test play session has previously been started with, not the scene that is currently running, i.e. the `current_scene` of the remote `SceneTree`, which may have been arrived at after one or several scene changes.

Well, yes, it doesn't respect any changes that you've made to the scene tree at runtime. If you've completely replaced your entire scene, the editor is not aware of that. That's an interesting use case, but I don't think anyone was considering that behavior before, in none of the previous discussions.

**Comment by YuriSizov** [comment]
> the proposal [godotengine/godot-proposals#6550](https://github.com/godotengine/godot-proposals/issues/6550) (which is to bring back the original behavior) exists specifically because they run into issues with scene dependencies with the new behavior.

> the proposal https://github.com/godotengine/godot-proposals/issues/6550 (which is to bring back the original behavior) exists specifically because they run into issues with scene dependencies with the new behavior.

Case in point, this proposal is also not about runtime changes. They are talking about running one scene, then editing a different scene, hitting F6 and getting the first scene restarted. Which is unexpected, but is a result of the fact that the F6 shortcut and the GUI button trigger the same action.

They are not talking about replacing their runtime scene tree using `change_scene_to` or similar. I'm pretty sure nobody but you is describing this case. Which is NOT to say your case is invalid. I'm just pointing out that you may be misunderstanding the cases of others, which do not in fact align with yours.

**Comment by h0lley** [comment]
I initially mentioned the idea that the function could respect runtime changes as an aside to salvage any usefulness of the current behavior (if it did that, then I guess I would see the use in it, as a SEPARATE function to the original one). I then brought it up again as that is how I'd interpret OP in #31515. "currently **running** scene" is not the specific scene the session has been started with, but, well, the scene that is currently running. so I think it's fair to say the current behavior has never been suggested by that user.

it is not my case however. my case is the same as the one of that proposal; to bring back the original significantly more intuitive behavior of F6.

**Comment by h0lley** [comment]
> Intuitive to you, intuition is individual, I personally expect the current behaviour, you can't make blanket statements based on intuition when it clearly differs, hence the change being made to begin with

I don't want to repeat myself but what behavior do you think takes more lines of text to explain?

> intuitve (Godot 3):
select the desired scene from the scene tab bar, press F6 to launch into that scene.

> unintuitive (current):
select the desired scene from the scene tab bar, press F6 to launch into that scene,
this played scene is now a special scene that is continuously played when launching with F6,
it may or may not still be the current scene (the scene you are currently looking at in the editor),
you have to quit test play first if you want to launch into a "new current scene".

granted, you could argue that in the scenario where the user wants to keep testing the same scene (let's call it "level"), and when they edit another scene (perhaps a dependency of level), having to tab back to the level scene before hitting F6 again is the extra step. so I can see why the proposal mentioned earlier suggests not to revert, but to bring back the original behavior as separate function (perhaps triggered with a modifier key). this workflow however requires advanced knowledge and I think it would be better for the original, much simpler behavior to be the default.

**Comment by Deozaan** [comment]
I agree that when I press F6, I want to run the scene I have currently selected/open in the editor, not reload whatever scene I might have left running in the window when I switched back to the editor.

I've internalized that F5 is to test the project starting from the main scene as configured in the project settings and F6 is to test the project starting from the currently selected scene. I've always understood the "reload" icon that appears there to be a shortcut for closing the debug instance and then clicking the play button again. More of a "relaunch" or "restart" rather than a "load that scene again" button.

I often forget to close the debug instance when I switch back to the editor to make some changes. I'm especially likely to do this if something isn't working the way I expected it to (e.g. a bug in my code) that I want to tweak real quick before a re-test. And sometimes I forget entirely that I've left the debug instance running and I work on a variety of scenes for quite a while before testing the project again. I don't think I would ever not be bewildered by F6 opening a different scene than the one I have right in front of me when pressing the "test the scene I have right in front of me" hotkey.

**Comment by hsandt** [comment]
Since the current behavior is apparently by design, and I have opened a proposal to add a separate button to avoid confusion, should we close this issue and redirect users to the proposal: https://github.com/godotengine/godot-proposals/issues/6550 so we can centralize the discussion (and the votes up on the first post)?

**Comment by Calinou** [comment]
Closing in favor of https://github.com/godotengine/godot-proposals/issues/6550.
